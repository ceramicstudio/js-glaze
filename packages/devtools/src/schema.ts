import { CIP88_REF_PREFIX } from '@glazed/constants'
import type { Schema } from '@glazed/types'

/** @internal */
export function getReference(schema: Schema): Array<string> | null {
  if (schema.$comment?.startsWith(CIP88_REF_PREFIX)) {
    const schemasString = schema.$comment.substr(CIP88_REF_PREFIX.length)
    if (schemasString.length) {
      const schemas = schemasString.split('|')
      schemas.sort()
      return schemas
    }
  }
  return null
}

/**
 * Recursively extract references to other schemas from a JSON schema arrays and objects
 *
 * @internal */
export function extractSchemaReferences(schema: Schema, path = ''): Record<string, Array<string>> {
  if (schema.type === 'string') {
    const refs = getReference(schema)
    return refs != null && refs.length > 0 ? { [path]: refs } : {}
  }
  if (schema.type === 'array') {
    return extractSchemaReferences(schema.items, path)
  }
  if (schema.type === 'object' && schema.properties != null) {
    // TODO: extract collection slice schema URL
    return Object.entries(schema.properties as Record<string, Schema>).reduce(
      (acc, [key, prop]) => {
        const propPath = path === '' ? key : `${path}.${key}`
        return Object.assign(acc, extractSchemaReferences(prop, propPath))
      },
      {} as Record<string, Array<string>>
    )
  }
  return {}
}
