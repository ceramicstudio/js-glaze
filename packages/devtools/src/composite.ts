import type { CeramicApi } from '@ceramicnetwork/common'
import type {
  CompositeViewsDefinition,
  EncodedCompositeDefinition,
  InternalCompositeDefinition,
  Model,
  ModelDefinition,
  RuntimeCompositeDefinition,
  StreamCommits,
} from '@glazed/types'
import type { GraphQLSchema } from 'graphql'
import cloneDeep from 'lodash-es/cloneDeep.js'
import merge from 'lodash-es/merge.js'
import createObjectHash from 'object-hash'

import { decodeSignedMap, encodeSignedMap } from './formats/json.js'
import { createRuntimeDefinition } from './formats/runtime.js'
import { compositeModelsAndCommonEmbedsFromGraphQLSchema } from './schema.js'

type StrictCompositeDefinition = Required<InternalCompositeDefinition>

function toStrictDefinition(definition: InternalCompositeDefinition): StrictCompositeDefinition {
  return {
    aliases: {},
    commonEmbeds: [],
    views: { account: {}, root: {}, models: {} },
    ...definition,
  }
}

function isSupportedVersion(supported: string, check: string): boolean {
  const [supportedMajor] = supported.split('.')
  const [checkMajor] = check.split('.')
  return supportedMajor === checkMajor
}

function assertSupportedVersion(supported: string, check: string): void {
  if (!isSupportedVersion(supported, check)) {
    throw new Error(`Unsupported Composite version ${check}, expected version ${supported}`)
  }
}

function assertModelsHaveCommits(
  models: Record<string, ModelDefinition>,
  commits: Record<string, StreamCommits>
): void {
  for (const id of Object.keys(models)) {
    if (commits[id] == null) {
      throw new Error(`Missing commits for model ${id}`)
    }
  }
}

export function setDefinitionAliases(
  definition: StrictCompositeDefinition,
  aliases: Record<string, string>,
  replace = false
): StrictCompositeDefinition {
  const existing = replace ? {} : definition.aliases
  definition.aliases = { ...existing, ...aliases }
  return definition
}

export function setDefinitionCommonEmbeds(
  definition: StrictCompositeDefinition,
  names: Iterable<string>,
  replace = false
): StrictCompositeDefinition {
  const existing = replace ? [] : definition.commonEmbeds
  definition.commonEmbeds = Array.from(new Set([...existing, ...names]))
  return definition
}

export function setDefinitionViews(
  definition: StrictCompositeDefinition,
  views: CompositeViewsDefinition,
  replace = false
): StrictCompositeDefinition {
  const existing = replace ? {} : definition.views
  definition.views = merge(existing, views)
  return definition
}

async function loadModelsFromCommits<Models = Record<string, StreamCommits>>(
  _ceramic: CeramicApi,
  modelsCommits: Models
): Promise<Record<keyof Models, ModelDefinition>> {
  const definitions = await Promise.all(
    Object.values(modelsCommits).map((_commits): Promise<Model> => {
      throw new Error('Not implemented')
      // return await Model.fromCommits(ceramic, commits)
    })
  )
  return Object.keys(modelsCommits).reduce((acc, id, index) => {
    const model = definitions[index]
    if (model == null) {
      throw new Error(`Missing Model for id ${id}`)
    }
    const modelID = model.id.toString()
    if (modelID !== id) {
      throw new Error(`Unexpected Model stream ID: expected ${id}, got ${modelID}`)
    }
    acc[modelID as keyof Models] = model.content
    return acc
  }, {} as Record<keyof Models, ModelDefinition>)
}

export type CompositeParams = {
  commits: Record<string, StreamCommits>
  definition: InternalCompositeDefinition
}

export type CompositeInput = Composite | CompositeParams
export type CompositeOptions = {
  aliases?: Record<string, string>
  commonEmbeds?: 'all' | 'none' | Iterable<string>
  views?: CompositeViewsDefinition
}

export type CreateParams = {
  ceramic: CeramicApi
  schema: string | GraphQLSchema
}

export type FromJSONParams = {
  ceramic: CeramicApi
  definition: EncodedCompositeDefinition
}

export type FromModelsParams = CompositeOptions & {
  ceramic: CeramicApi
  models: Array<string>
}

export class Composite {
  static VERSION = '1.0'

  static create(params: CreateParams): Promise<Composite> {
    // TODO: convert the schema to models, create the models on the Ceramice node, load their commits

    const { models, commonEmbeds } = compositeModelsAndCommonEmbedsFromGraphQLSchema(params.schema)
    console.log('MODELS', models)
    console.log('COMMON EMBEDS', commonEmbeds)

    throw new Error('Not implemented')
  }

  static from(composites: Iterable<CompositeInput>, options?: CompositeOptions): Composite {
    const [first, ...rest] = composites
    if (first == null) {
      throw new Error('Missing composites to compose')
    }
    const composite = first instanceof Composite ? first.clone() : new Composite(first)
    return composite.merge(rest, options)
  }

  static async fromJSON(params: FromJSONParams): Promise<Composite> {
    const { models, ...definition } = params.definition
    const commits = decodeSignedMap(models)
    return new Composite({
      commits,
      definition: {
        ...definition,
        models: await loadModelsFromCommits(params.ceramic, commits),
      },
    })
  }

  static fromModels(_params: FromModelsParams): Promise<Composite> {
    // TODO: load all models definitions and commits
    throw new Error('Not implemented')
  }

  #commits: Record<string, StreamCommits>
  #definition: StrictCompositeDefinition
  #hash: string | undefined

  constructor(params: CompositeParams) {
    assertSupportedVersion(Composite.VERSION, params.definition.version)
    assertModelsHaveCommits(params.definition.models, params.commits)
    this.#commits = cloneDeep(params.commits)
    this.#definition = toStrictDefinition(cloneDeep(params.definition))
  }

  get hash(): string {
    if (this.#hash == null) {
      this.#hash = createObjectHash(this.#definition)
    }
    return this.#hash
  }

  clone(): Composite {
    return new Composite(this.toParams())
  }

  copy(models: Array<string>): Composite {
    if (models.length === 0) {
      throw new Error('Missing models to copy')
    }

    const { commits, definition } = this.toParams()
    const def = toStrictDefinition(definition)

    const nameIDs = Object.entries(def.models).reduce((acc, [id, model]) => {
      acc[model.name] = id
      return acc
    }, {} as Record<string, string>)
    const aliasIDs = Object.entries(def.aliases).reduce((acc, [id, alias]) => {
      acc[alias] = id
      return acc
    }, {} as Record<string, string>)

    const nextCommits: Record<string, StreamCommits> = {}
    const nextModels: Record<string, ModelDefinition> = {}
    const nextAliases: Record<string, string> = {}
    const nextViews: CompositeViewsDefinition = { account: {}, root: {}, models: {} }

    for (const model of models) {
      const id = aliasIDs[model] ?? nameIDs[model] ?? model
      if (def.models[id] == null) {
        throw new Error(`Model not found: ${model}`)
      }

      nextCommits[id] = commits[id]
      nextModels[id] = def.models[id]
      if (def.aliases[id] != null) {
        nextAliases[id] = def.aliases[id]
      }
      if (def.views.models[id] != null) {
        // TODO: check relations to other models, ensure there's no missing model in the subset
        nextViews.models[id] = def.views.models[id]
      }
      // TODO: account and root views
    }

    return new Composite({
      commits: nextCommits,
      definition: {
        version: def.version,
        commonEmbeds: def.commonEmbeds,
        models: nextModels,
        aliases: nextAliases,
        views: nextViews,
      },
    })
  }

  equals(other: CompositeInput): boolean {
    const otherHash =
      other instanceof Composite
        ? other.hash
        : createObjectHash(toStrictDefinition(other.definition))
    return this.hash === otherHash
  }

  merge(other: CompositeInput | Array<CompositeInput>, options: CompositeOptions = {}): Composite {
    const nextParams = this.toParams()
    const nextDefinition = toStrictDefinition(nextParams.definition)
    const collectedEmbeds = new Set<string>()

    for (const composite of Array.isArray(other) ? other : [other]) {
      const { commits, definition } =
        composite instanceof Composite ? composite.toParams() : composite
      assertSupportedVersion(nextDefinition.version, definition.version)
      assertModelsHaveCommits(definition.models, commits)

      const def = toStrictDefinition(definition)
      Object.assign(nextParams.commits, commits)
      Object.assign(nextDefinition.models, definition.models)
      Object.assign(nextDefinition.aliases, def.aliases)
      merge(nextDefinition.views, def.views)
      for (const name of def.commonEmbeds) {
        collectedEmbeds.add(name)
      }
    }

    if (options.aliases != null) {
      setDefinitionAliases(nextDefinition, options.aliases)
    }
    const commonEmbeds = options.commonEmbeds ?? 'none'
    if (commonEmbeds === 'all') {
      setDefinitionCommonEmbeds(nextDefinition, collectedEmbeds)
    } else if (Array.isArray(commonEmbeds)) {
      setDefinitionCommonEmbeds(nextDefinition, commonEmbeds, true)
    }
    if (options.views != null) {
      setDefinitionViews(nextDefinition, options.views)
    }

    return new Composite({ ...nextParams, definition: nextDefinition })
  }

  setAliases(aliases: Record<string, string>, replace = false): Composite {
    const params = this.toParams()
    const definition = setDefinitionAliases(toStrictDefinition(params.definition), aliases, replace)
    return new Composite({ ...params, definition })
  }

  setCommonEmbeds(names: Iterable<string>, replace = false): Composite {
    const params = this.toParams()
    const definition = setDefinitionCommonEmbeds(
      toStrictDefinition(params.definition),
      names,
      replace
    )
    return new Composite({ ...params, definition })
  }

  setViews(views: CompositeViewsDefinition, replace = false): Composite {
    const params = this.toParams()
    const definition = setDefinitionViews(toStrictDefinition(params.definition), views, replace)
    return new Composite({ ...params, definition })
  }

  toJSON(): EncodedCompositeDefinition {
    return {
      version: this.#definition.version,
      models: encodeSignedMap(this.#commits),
      aliases: this.#definition.aliases,
      views: this.#definition.views,
    }
  }

  toParams(): CompositeParams {
    return {
      commits: cloneDeep(this.#commits),
      definition: cloneDeep(this.#definition),
    }
  }

  toRuntime(): RuntimeCompositeDefinition {
    return createRuntimeDefinition(this.#definition)
  }
}
