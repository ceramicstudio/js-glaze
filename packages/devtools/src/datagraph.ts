export class DataGraph {
  // Deterministic account to single record (IDX)
  async getRecord(type: string, did?: string): Promise<any> {}
  async setRecord(type: string, content: any): Promise<any> {}
  async removeRecord(type: string): Promise<any> {}

  // Account to multiple records
  async listCollectionRecords(type: string, did?: string): Promise<any> {}
  async addCollectionRecord(type: string, content: any): Promise<any> {}
  async removeCollectionRecord(id: string): Promise<any> {}

  // Deterministic account links
  async getAccountLink(type: string, did: string): Promise<any> {}
  async setAccountLink(type: string, did: string, content?: any): Promise<any> {}
  async removeAccountLink(type: string, did: string): Promise<any> {}

  // Account links query
  async listAccountLinksFrom(type: string, did?: string): Promise<any> {}
  async listAccountLinksTo(type: string, did?: string): Promise<any> {}
  async listReciprocalAccountLinks(type: string, did?: string): Promise<any> {}
  async listAccountLinksRecords(
    linkType: string,
    collectionType: string,
    did?: string
  ): Promise<any> {}

  // Records associated to stream -- should these use the CollectionRecords APIs?
  // Need to define query parameters for collections
  async listChildrenRecords(streamID: string, type: string): Promise<any> {}
  async addChildRecord(streamID: string, type: string, content: any): Promise<any> {}
  // Should this simply be removeCollectionRecord?
  async removeChildRecord(streamID: string, type: string, id: string): Promise<any> {}

  // TODO: getCollection with all basic profiles
}
