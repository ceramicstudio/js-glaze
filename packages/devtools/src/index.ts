/**
 * ```sh
 * npm install --dev @glazed/devtools
 * ```
 *
 * @module devtools
 */

export * from './datamodel'
export * from './encoding'
export * from './graphql'
export * from './publishing'
export * from './schema'
export * from './utils'
export * from './validation'
