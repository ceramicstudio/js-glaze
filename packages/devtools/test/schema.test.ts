/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { compositeModelsAndCommonEmbedsFromGraphQLSchema } from '../src/schema'
import { compositeSchemaWithProfiles } from './exampleSchemas/compositeSchemaWithProfiles.schema'
import { graphQLSchemaWithoutModels } from './exampleSchemas/graphQLSchemaWithoutModels.schema'
import ajv from 'ajv/dist/2020'

describe('schema', () => {
  it("compositeModelsAndCommonEmbedsFromGraphQLSchema throws when there's no top-level model object", () => {
    expect(() => {
      compositeModelsAndCommonEmbedsFromGraphQLSchema(graphQLSchemaWithoutModels)
    }).toThrow('No models found in Composite Definition Schema')
  })

  it("compositeModelsAndCommonEmbedsFromGraphQLSchema doesn't parse unions", () => {
    expect(() => {
      compositeModelsAndCommonEmbedsFromGraphQLSchema(`
      union IntOrString = Int | String

      type ModelWithUnionProp @model(
        accountRelation: LINK,
        description: "Test model with a property that is a union of string and int"
      ) {
        intOrStringValue: IntOrString
      }
      `)
    }).toThrow('GraphQL unions are not supported')
  })

  it("compositeModelsAndCommonEmbedsFromGraphQLSchema doesn't parse interfaces", () => {
    expect(() => {
      compositeModelsAndCommonEmbedsFromGraphQLSchema(`
      interface GenericProfile {
        name: String @length(max: 150)
      }
       
      type SocialProfile implements GenericProfile @model(
        accountRelation: LINK,
        description: "A model to store properties that accounts would like to share on social media"
      ) {
        description: String @length(max: 420)
        emoji: String @length(max: 2)
        url: String @length(max: 240)
      }
      `)
    }).toThrow('GraphQL interfaces are not supported')
  })

  it('compositeModelsAndCommonEmbedsFromGraphQLSchema creates an InternalCompositeDefinition for profiles from schema', () => {
    expect(
      compositeModelsAndCommonEmbedsFromGraphQLSchema(compositeSchemaWithProfiles)
    ).toMatchSnapshot()
  })

  it('compositeModelsAndCommonEmbedsFromGraphQLSchema creates models whose schemas conform to JSON Schema standard', () => {
    const compositeDefinition = compositeModelsAndCommonEmbedsFromGraphQLSchema(
      compositeSchemaWithProfiles
    )
    const [genericProfile, socialProfile, personalProfile] = compositeDefinition.models

    const validator = new ajv()
    expect(validator.validateSchema(genericProfile.schema, true)).toBe(true)
    expect(validator.validateSchema(socialProfile.schema, true)).toBe(true)
    expect(validator.validateSchema(personalProfile.schema, true)).toBe(true)
  })

  it('DID scalar is supported and properly converted to ICD', () => {
    expect(
      compositeModelsAndCommonEmbedsFromGraphQLSchema(`
      type ModelWithDIDProp @model(
        accountRelation: LINK,
        description: "Test model with DID properties"
      ) {
        didValue: DID
        requiredDidValue: DID!
      }
      `)
    ).toMatchObject({
      models: [
        {
          name: 'ModelWithDIDProp',
          accountRelation: 'link',
          schema: {
            $schema: 'https://json-schema.org/draft/2020-12/schema',
            type: 'object',
            properties: {
              didValue: {
                type: 'string',
                title: 'GraphQLDID',
                pattern:
                  "/^did:[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+:[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+$/",
                maxLength: 80,
              },
              requiredDidValue: {
                type: 'string',
                title: 'GraphQLDID',
                pattern:
                  "/^did:[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+:[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+$/",
                maxLength: 80,
              },
            },
            additionalProperties: false,
            required: ['requiredDidValue'],
          },
        },
      ],
    })
  })

  it('StreamReference scalar is supported and properly converted to ICD', () => {
    expect(
      compositeModelsAndCommonEmbedsFromGraphQLSchema(`
      type ModelWithStreamReferenceProp @model(
        accountRelation: LINK,
        description: "Test model with stream reference properties"
      ) {
        streamReferenceValue: StreamReference
        requiredStreamReferenceValue: StreamReference!
      }
      `)
    ).toMatchObject({
      models: [
        {
          name: 'ModelWithStreamReferenceProp',
          accountRelation: 'link',
          schema: {
            $schema: 'https://json-schema.org/draft/2020-12/schema',
            type: 'object',
            properties: {
              streamReferenceValue: {
                type: 'string',
                title: 'CeramicStreamReference',
                pattern: '<TBD>',
                maxLength: 80,
              },
              requiredStreamReferenceValue: {
                type: 'string',
                title: 'CeramicStreamReference',
                pattern: '<TBD>',
                maxLength: 80,
              },
            },
            additionalProperties: false,
            required: ['requiredStreamReferenceValue'],
          },
        },
      ],
    })
  })

  it('Boolean scalar is supported and properly converted to ICD', () => {
    expect(
      compositeModelsAndCommonEmbedsFromGraphQLSchema(`
      type ModelWithBooleanProp @model(
        accountRelation: LINK,
        description: "Test model with boolean properties"
      ) {
        booleanValue: Boolean
        requiredBooleanValue: Boolean!
      }
      `)
    ).toMatchObject({
      models: [
        {
          name: 'ModelWithBooleanProp',
          accountRelation: 'link',
          schema: {
            $schema: 'https://json-schema.org/draft/2020-12/schema',
            type: 'object',
            properties: {
              booleanValue: {
                type: 'boolean',
              },
              requiredBooleanValue: {
                type: 'boolean',
              },
            },
            additionalProperties: false,
            required: ['requiredBooleanValue'],
          },
        },
      ],
    })
  })

  it('Int scalar is supported and properly converted to ICD', () => {
    expect(
      compositeModelsAndCommonEmbedsFromGraphQLSchema(`
      type ModelWithIntProp @model(
        accountRelation: LINK,
        description: "Test model with int properties"
      ) {
        intValue: Int
        requiredIntValue: Int!
      }
      `)
    ).toMatchObject({
      models: [
        {
          name: 'ModelWithIntProp',
          accountRelation: 'link',
          schema: {
            $schema: 'https://json-schema.org/draft/2020-12/schema',
            type: 'object',
            properties: {
              intValue: {
                type: 'integer',
              },
              requiredIntValue: {
                type: 'integer',
              },
            },
            additionalProperties: false,
            required: ['requiredIntValue'],
          },
        },
      ],
    })
  })

  it('Float scalar is supported and properly converted to ICD', () => {
    expect(
      compositeModelsAndCommonEmbedsFromGraphQLSchema(`
      type ModelWithFloatProp @model(
        accountRelation: LINK,
        description: "Test model with float properties"
      ) {
        floatValue: Float
        requiredFloatValue: Float!
      }
      `)
    ).toMatchObject({
      models: [
        {
          name: 'ModelWithFloatProp',
          accountRelation: 'link',
          schema: {
            $schema: 'https://json-schema.org/draft/2020-12/schema',
            type: 'object',
            properties: {
              floatValue: {
                type: 'number',
              },
              requiredFloatValue: {
                type: 'number',
              },
            },
            additionalProperties: false,
            required: ['requiredFloatValue'],
          },
        },
      ],
    })
  })

  it('String scalar is supported and properly converted to ICD', () => {
    expect(
      compositeModelsAndCommonEmbedsFromGraphQLSchema(`
      type ModelWithStringProp @model(
        accountRelation: LINK,
        description: "Test model with string properties"
      ) {
        stringValue: String
        requiredStringValue: String!
      }
      `)
    ).toMatchObject({
      models: [
        {
          name: 'ModelWithStringProp',
          accountRelation: 'link',
          schema: {
            $schema: 'https://json-schema.org/draft/2020-12/schema',
            type: 'object',
            properties: {
              stringValue: {
                type: 'string',
              },
              requiredStringValue: {
                type: 'string',
              },
            },
            additionalProperties: false,
            required: ['requiredStringValue'],
          },
        },
      ],
    })
  })

  it('ID scalar is supported and properly converted to ICD', () => {
    expect(
      compositeModelsAndCommonEmbedsFromGraphQLSchema(`
      type ModelWithIDProp @model(
        accountRelation: LINK,
        description: "Test model with GraphQL ID property"
      ) {
        idValue: ID
        requiredIdValue: ID!
      }
      `)
    ).toMatchObject({
      models: [
        {
          name: 'ModelWithIDProp',
          accountRelation: 'link',
          schema: {
            $schema: 'https://json-schema.org/draft/2020-12/schema',
            type: 'object',
            properties: {
              idValue: {
                type: 'string',
                title: 'GraphQLID',
              },
              requiredIdValue: {
                type: 'string',
                title: 'GraphQLID',
              },
            },
            additionalProperties: false,
            required: ['requiredIdValue'],
          },
        },
      ],
    })
  })

  it('Arrays are supported and properly converted to ICD', () => {
    expect(
      compositeModelsAndCommonEmbedsFromGraphQLSchema(`
      type ModelWithArrayProp @model(
        accountRelation: LINK,
        description: "Test model with GraphQL ID property"
      ) {
        arrayValue: [Int]
        requiredArrayValue: [Int]!
      }
      `)
    ).toMatchObject({
      models: [
        {
          name: 'ModelWithArrayProp',
          accountRelation: 'link',
          schema: {
            $schema: 'https://json-schema.org/draft/2020-12/schema',
            type: 'object',
            properties: {
              arrayValue: {
                type: 'array',
                items: {
                  type: 'integer',
                },
              },
              requiredArrayValue: {
                type: 'array',
                items: {
                  type: 'integer',
                },
              },
            },
            additionalProperties: false,
            required: ['requiredArrayValue'],
          },
        },
      ],
    })
  })

  it('@length(min: Int, max: Int) directive is supported for strings and properly converted to ICD', () => {
    expect(
      compositeModelsAndCommonEmbedsFromGraphQLSchema(`
      type ModelWithStringProp @model(
        accountRelation: LINK,
        description: "Test model with a constrained string property"
      ) {
        stringValue: String @length(min: 1, max: 140)
      }
      `)
    ).toMatchObject({
      models: [
        {
          name: 'ModelWithStringProp',
          accountRelation: 'link',
          schema: {
            $schema: 'https://json-schema.org/draft/2020-12/schema',
            type: 'object',
            properties: {
              stringValue: {
                type: 'string',
                minLength: 1,
                maxLength: 140,
              },
            },
            additionalProperties: false,
          },
        },
      ],
    })
  })

  it('@intRange(min: Int, max: Int) can be applied to Int, Int! or [Int] properties', () => {
    expect(() => {
      compositeModelsAndCommonEmbedsFromGraphQLSchema(`
      type ModelWithStringProp @model(
        accountRelation: LINK,
        description: "Test model with incorrectly constrained string property"
      ) {
        intValue: String @intRange(min: 1)
      }
      `)
    }).toThrow('@intRange can only be applied to integers or arrays of integers')
  })

  it('@floatRange(min: Int, max: Int) can be applied to Float, Float! or [Float] properties', () => {
    expect(() => {
      compositeModelsAndCommonEmbedsFromGraphQLSchema(`
      type ModelWithIntProp @model(
        accountRelation: LINK,
        description: "Test model with incorrectly constrained int property"
      ) {
        intValue: Int @floatRange(max: 1)
      }
      `)
    }).toThrow('@floatRange can only be applied to floats or arrays of floats')
  })

  it('@length(min: Int, max: Int) can be applied to strings or arrays of strings', () => {
    expect(() => {
      compositeModelsAndCommonEmbedsFromGraphQLSchema(`
      type ModelWithArrayProp @model(
        accountRelation: LINK,
        description: "Test model with incorrectly constrained array property"
      ) {
        intValue: [Int] @length(min:10, max: 140)
      }
      `)
    }).toThrow('@length can only be applied to strings or arrays of strings')
  })

  it('@arrayLength(min: Int, max: Int) can be applied to arrays', () => {
    expect(() => {
      compositeModelsAndCommonEmbedsFromGraphQLSchema(`
      type ModelWithStringProp @model(
        accountRelation: LINK,
        description: "Test model with incorrectly constrained strings property"
      ) {
        intValue: String @arrayLength(max: 140)
      }
      `)
    }).toThrow('@arrayLength can only be applied to arrays')
  })

  it('@arrayLength(min: Int, max: Int) directive is supported for arrays and properly converted to ICD', () => {
    expect(
      compositeModelsAndCommonEmbedsFromGraphQLSchema(`
      type ModelWithArrayProp @model(
        accountRelation: LINK,
        description: "Test model with a constrained array property"
      ) {
        arrayValue: [String] @arrayLength(min: 10, max: 15)
      }
      `)
    ).toMatchObject({
      models: [
        {
          name: 'ModelWithArrayProp',
          accountRelation: 'link',
          schema: {
            $schema: 'https://json-schema.org/draft/2020-12/schema',
            type: 'object',
            properties: {
              arrayValue: {
                type: 'array',
                minItems: 10,
                maxItems: 15,
                items: {
                  type: 'string',
                },
              },
            },
            additionalProperties: false,
          },
        },
      ],
    })
  })

  it('@length(min: Int, max: Int) directive is supported for arrays of strings and properly converted to ICD', () => {
    expect(
      compositeModelsAndCommonEmbedsFromGraphQLSchema(`
      type ModelWithArrayProp @model(
        accountRelation: LINK,
        description: "Test model with an array property with constrained items"
      ) {
        arrayValue: [String] @length(min: 4, max: 440)
      }
      `)
    ).toMatchObject({
      models: [
        {
          name: 'ModelWithArrayProp',
          accountRelation: 'link',
          schema: {
            $schema: 'https://json-schema.org/draft/2020-12/schema',
            type: 'object',
            properties: {
              arrayValue: {
                type: 'array',
                items: {
                  type: 'string',
                  minLength: 4,
                  maxLength: 440,
                },
              },
            },
            additionalProperties: false,
          },
        },
      ],
    })
  })

  it('@intRange(min: Int, max: Int) directive is supported and properly converted to ICD', () => {
    expect(
      compositeModelsAndCommonEmbedsFromGraphQLSchema(`
      type ModelWithIntProp @model(
        accountRelation: LINK,
        description: "Test model with a constreained int property"
      ) {
        intValue: Int @intRange(min: 5, max: 10)
      }
      `)
    ).toMatchObject({
      models: [
        {
          name: 'ModelWithIntProp',
          accountRelation: 'link',
          schema: {
            $schema: 'https://json-schema.org/draft/2020-12/schema',
            type: 'object',
            properties: {
              intValue: {
                type: 'integer',
                minimum: 5,
                maximum: 10,
              },
            },
            additionalProperties: false,
          },
        },
      ],
    })
  })

  it('@floatRange(min: Float, max: Float) directive is supported and properly converted to ICD', () => {
    expect(
      compositeModelsAndCommonEmbedsFromGraphQLSchema(`
      type ModelWithFloatProp @model(
        accountRelation: LINK,
        description: "Test model with a constrained float property"
      ) {
        floatValue: Float @floatRange(min: 5.0, max: 10.0)
      }
      `)
    ).toMatchObject({
      models: [
        {
          name: 'ModelWithFloatProp',
          accountRelation: 'link',
          schema: {
            $schema: 'https://json-schema.org/draft/2020-12/schema',
            type: 'object',
            properties: {
              floatValue: {
                type: 'number',
                minimum: 5.0,
                maximum: 10.0,
              },
            },
            additionalProperties: false,
          },
        },
      ],
    })
  })
})
