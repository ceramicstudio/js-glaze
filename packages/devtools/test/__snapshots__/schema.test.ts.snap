// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema parseCompositeSchema creates an InternalCompositeDefinition for profiles from schema 1`] = `
Object {
  "commonEmbeds": Array [
    "ImageSources",
    "ImageMetadata",
  ],
  "models": Array [
    Object {
      "accountRelation": "single",
      "description": "A model to store common profile-related properties",
      "name": "GenericProfile",
      "schema": Object {
        "$defs": Object {
          "ImageMetadata": Object {
            "additionalProperties": false,
            "properties": Object {
              "height": Object {
                "minimum": 1,
                "type": "integer",
              },
              "mimeType": Object {
                "maxLength": 50,
                "minLength": 0,
                "type": "string",
              },
              "size": Object {
                "minimum": 1,
                "type": "integer",
              },
              "src": Object {
                "maxLength": 150,
                "minLength": 0,
                "type": "string",
              },
              "width": Object {
                "minimum": 1,
                "type": "integer",
              },
            },
            "required": Array [
              "src",
              "mimeType",
              "width",
              "height",
            ],
            "title": "ImageMetadata",
            "type": "object",
          },
          "ImageSources": Object {
            "additionalProperties": false,
            "properties": Object {
              "alternatives": Object {
                "items": Object {
                  "$ref": "#/$defs/ImageMetadata",
                },
                "maxItems": 20,
                "type": "array",
              },
              "original": Object {
                "$ref": "#/$defs/ImageMetadata",
              },
            },
            "required": Array [
              "original",
            ],
            "title": "ImageSources",
            "type": "object",
          },
        },
        "$schema": "https://json-schema.org/draft/2020-12/schema",
        "additionalProperties": false,
        "properties": Object {
          "image": Object {
            "$ref": "#/$defs/ImageSources",
          },
          "name": Object {
            "maxLength": 150,
            "minLength": 0,
            "type": "string",
          },
        },
        "type": "object",
      },
    },
    Object {
      "accountRelation": "single",
      "description": "A model to store properties that accounts would like to share on social media",
      "name": "SocialProfile",
      "schema": Object {
        "$defs": Object {
          "ImageMetadata": Object {
            "additionalProperties": false,
            "properties": Object {
              "height": Object {
                "minimum": 1,
                "type": "integer",
              },
              "mimeType": Object {
                "maxLength": 50,
                "minLength": 0,
                "type": "string",
              },
              "size": Object {
                "minimum": 1,
                "type": "integer",
              },
              "src": Object {
                "maxLength": 150,
                "minLength": 0,
                "type": "string",
              },
              "width": Object {
                "minimum": 1,
                "type": "integer",
              },
            },
            "required": Array [
              "src",
              "mimeType",
              "width",
              "height",
            ],
            "title": "ImageMetadata",
            "type": "object",
          },
          "ImageSources": Object {
            "additionalProperties": false,
            "properties": Object {
              "alternatives": Object {
                "items": Object {
                  "$ref": "#/$defs/ImageMetadata",
                },
                "maxItems": 20,
                "type": "array",
              },
              "original": Object {
                "$ref": "#/$defs/ImageMetadata",
              },
            },
            "required": Array [
              "original",
            ],
            "title": "ImageSources",
            "type": "object",
          },
        },
        "$schema": "https://json-schema.org/draft/2020-12/schema",
        "additionalProperties": false,
        "properties": Object {
          "background": Object {
            "$ref": "#/$defs/ImageSources",
          },
          "description": Object {
            "maxLength": 420,
            "minLength": 0,
            "type": "string",
          },
          "emoji": Object {
            "maxLength": 2,
            "minLength": 0,
            "type": "string",
          },
          "url": Object {
            "maxLength": 240,
            "minLength": 0,
            "type": "string",
          },
        },
        "type": "object",
      },
    },
    Object {
      "accountRelation": "single",
      "description": "A model to store accounts' personal data",
      "name": "PersonProfile",
      "schema": Object {
        "$schema": "https://json-schema.org/draft/2020-12/schema",
        "additionalProperties": false,
        "properties": Object {
          "affiliations": Object {
            "items": Object {
              "maxLength": 140,
              "minLength": 0,
              "type": "string",
            },
            "maxItems": 20,
            "type": "array",
          },
          "birthDate": Object {
            "maxLength": 10,
            "minLength": 0,
            "type": "string",
          },
          "gender": Object {
            "maxLength": 42,
            "minLength": 0,
            "type": "string",
          },
          "homeLocation": Object {
            "maxLength": 140,
            "minLength": 0,
            "type": "string",
          },
          "nationalities": Object {
            "items": Object {
              "maxLength": 2,
              "minLength": 0,
              "type": "string",
            },
            "maxItems": 5,
            "type": "array",
          },
          "residenceCountry": Object {
            "maxLength": 2,
            "minLength": 0,
            "type": "string",
          },
        },
        "type": "object",
      },
    },
  ],
}
`;
